notebooks/03-query/16-export-msq-query.ipynb

```diff
@@ -0,0 +1,426 @@
+    "  \"isRobot\",\n",
+    "  \"channel\",\n",
+    "  \"added\",\n",
+    "  \"user\",\n",
+    "  \"deleted\"\n",
+    "FROM \"ext\"\n",
+    "PARTITIONED BY DAY\n",
+    "'''\n",
+    "display_client.run_task(sql)\n",
+    "sql_client.wait_until_ready('example-wiki-export')\n",
+    "display_client.table('example-wiki-export')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "36ad757a-957e-465a-a284-855c14e3e8fd",
+   "metadata": {},
+   "source": [
+    "## Exporting data\n",
+    "\n",
+    "Exporting data uses MSQ `INSERT` to an `EXTERN` table function to define the output location and format of the data.\n",
+    "\n",
+    "The SQL statement takes on the form:\n",
+    "```\n",
+    "INSERT INTO\n",
+    "  EXTERN(<destination function>)\n",
+    "AS CSV\n",
+    "SELECT\n",
+    "  ...\n",
+    "```\n",
+    "\n",
+    "- destination function defines the output location\n",
+    "- AS CSV is the format of the exported file(s), new formats are expected as this experimental function evolves\n",
+    "- SELECT statement can be any MSQ query including transformations, joins and aggregations\n",
+    "\n",
+    "In this learning environment the `druid_export_storage_baseDir` property has been set to `/opt/shared/exports` in the docker compose \"environment\" file.\n",
+    "The volume `/opt/shared` is accessible to all Druid processes as well as the Jupyter Labs container.\n",
+    "\n",
+    "The following example is a simple export of all data in the `example-wiki-export` datasource.\n",
+    "The output folder must exist and be empty. The following cell removes the folder if it exists and then creates the folder in the /opt/shared/exports:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "a474e38b-07aa-4f7e-9275-2a2cd18ac0fb",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "!rm -rf /opt/shared/exports/example-wiki-export\n",
+    "!mkdir -p /opt/shared/exports/example-wiki-export"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "96a9b601-dfa6-48a3-a00b-e710a812d2e2",
+   "metadata": {},
+   "source": [
+    "The next cell executes the the export:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "0b36cb33-1e8a-4284-966c-a3937c8a33a4",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "sql='''\n",
+    "    INSERT INTO \n",
+    "       EXTERN ( local(exportPath => '/opt/shared/exports/example-wiki-export') )\n",
+    "       AS CSV\n",
+    "    SELECT\n",
+    "      *\n",
+    "    FROM \"example-wiki-export\" \n",
+    "'''\n",
+    "display_client.run_task(sql)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f9571de0-eac8-40f0-8970-f137aec9e767",
+   "metadata": {},
+   "source": [
+    "Take a look at the file this command created, notice that the name contains the query id, the worker task that created the file and a partition number, each MSQ worker creates one or more partitions:"
```

Comment:
Run the following command to list the contents of the exports folder.

The filename contains:

* query Id.
* worker task Id.
* a partition number.

The worker task Id references the MSQ worker that created the final output.

The partition number is X, Y, Z and multiple blah blah.

Reply:
done